Container for trivially-copyable objects with tree output support

shahor02 opened the pull request:
Hi, @ktf, @mkrzewic , this is the vector-like contrainer for copyable objects T  which can be sent as and reinterpteted as char array and equipped with functionality to write itself as a vector<T> branch in the tree.
I see it as a prototype for the containers used to exchange reconstructed info between different components.
The persistency (either via binary, root file or TTree) is meant to allow their usage also for standalone tests etc. 
The Detectors/Base/test/testContVec.cxx gives an idea of main use cases. Apart from serving as unit-test, it can be also run as a root macro (after uncommenting  its 1st \#define )

shahor02 committed:
Container for trivially-copyable objects with tree output support

alibuild commented:
Error while checking build/O2/o2 for 0c601285c84e8d65841b7cbb9a95183893e2d675:
```
sw/BUILD/O2-latest/log
+ perl -p -i -e 's|/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/dev/0|/mnt/mesos/sandbox/sandbox/O2|' compile_commands.json
+ ln -sf /mnt/mesos/sandbox/sandbox/sw/BUILD/b3b37fa96d85a0e858b2e6ac9dda2e680aec17de/O2/compile_commands.json /mnt/mesos/sandbox/sandbox/O2/compile_commands.json
+ [[ -n 1 ]]
+ make test
Running tests...
Test project /mnt/mesos/sandbox/sandbox/sw/BUILD/b3b37fa96d85a0e858b2e6ac9dda2e680aec17de/O2
      Start  1: testMessageList
 1/17 Test  #1: testMessageList ..................   Passed    0.00 sec
      Start  2: testDataHeader
 2/17 Test  #2: testDataHeader ...................   Passed    0.57 sec
      Start  3: testTimeStamp
 3/17 Test  #3: testTimeStamp ....................   Passed    0.43 sec
      Start  4: testBasicHits
 4/17 Test  #4: testBasicHits ....................   Passed    0.48 sec
      Start  5: TimeFrameTest
 5/17 Test  #5: TimeFrameTest ....................   Passed    0.53 sec
      Start  6: testContVec
 6/17 Test  #6: testContVec ......................   Passed    0.74 sec
      Start  7: testTPCBase
 7/17 Test  #7: testTPCBase ......................   Passed    0.43 sec
      Start  8: testTPCCalDet
 8/17 Test  #8: testTPCCalDet ....................   Passed    0.65 sec
      Start  9: testTPCMapper
 9/17 Test  #9: testTPCMapper ....................   Passed    0.41 sec
      Start 10: testTPCSyncPatternMonitor
10/17 Test #10: testTPCSyncPatternMonitor ........   Passed    0.46 sec
      Start 11: testTPCAdcClockMonitor
11/17 Test #11: testTPCAdcClockMonitor ...........   Passed    0.47 sec
      Start 12: testTPCElectronTransport
12/17 Test #12: testTPCElectronTransport .........   Passed    0.64 sec
      Start 13: testTPCGEMAmplification
13/17 Test #13: testTPCGEMAmplification ..........   Passed   24.83 sec
      Start 14: testTPCSimulation
14/17 Test #14: testTPCSimulation ................   Passed    0.55 sec
      Start 15: run_flp2epn_distributed
15/17 Test #15: run_flp2epn_distributed ..........***Failed  Required regular expression not found.Regex=[acknowledged after
]  5.21 sec
      Start 16: testExampleModule1
16/17 Test #16: testExampleModule1 ...............   Passed    0.02 sec
      Start 17: O2MessageMonitorTest
17/17 Test #17: O2MessageMonitorTest .............   Passed    0.53 sec

94% tests passed, 1 tests failed out of 17

Total Test time (real) =  36.97 sec

The following tests FAILED:
	 15 - run_flp2epn_distributed (Failed)
Errors while running CTest
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/0c601285c84e8d65841b7cbb9a95183893e2d675/fullLog.txt).

alibuild commented:
Error while checking build/o2checkcode/o2-daq for 0c601285c84e8d65841b7cbb9a95183893e2d675:
```
sw/BUILD/o2checkcode-latest/log
+ grep -v /G__ error-log.txt
+ grep -v .pb.cc
+ grep ' error:'
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0/Detectors/Base/test/testContVec.cxx:96:16: error: use nullptr [modernize-use-nullptr]
/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0/Detectors/Base/test/testContVec.cxx:171:46: error: use nullptr [modernize-use-nullptr]
+ exit 1sw/BUILD/O2-latest/log
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/QCViewer/ViewerDevice.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/runQCViewerDevice
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/runQCViewerDevice" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libQCViewer.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libQCMetricsExtractor.so
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libQCMetricsExtractor.so" to ""
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/QCMetricsExtractor
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/QCMetricsExtractor/MetricsExtractor.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/runQCMetricsExtractor
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/runQCMetricsExtractor" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libQCMetricsExtractor.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so" to ""
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/DataFlow
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/DataFlow/SubframeBuilderDevice.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/DataFlow/TimeframeValidatorDevice.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/DataFlow/EPNReceiverDevice.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/DataFlow/HeartbeatSampler.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/DataFlow/FLPSenderDevice.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/heartbeatSampler
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/heartbeatSampler" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/SubframeBuilderDevice
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/SubframeBuilderDevice" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/TimeframeValidatorDevice
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/TimeframeValidatorDevice" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/EPNReceiverDevice
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/EPNReceiverDevice" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/FLPSenderDevice
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/FLPSenderDevice" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libDataFlow.so
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libPublishers.so
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libPublishers.so" to ""
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/Publishers
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/include/Publishers/DataPublisherDevice.h
-- Installing: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/DataPublisherDevice
-- Set runtime path of "/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/bin/DataPublisherDevice" to ""
-- Up-to-date: /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/lib/libPublishers.so
+ mkdir -p /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/etc/modulefiles
+ rsync -a --delete etc/modulefiles/ /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1/etc/modulefiles
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/0c601285c84e8d65841b7cbb9a95183893e2d675/fullLog.txt).

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+namespace o2 {
+namespace Base {
+
+using sizeType = int;
```
perhaps use size_t (or if this is supposed to be transported more explicitly int64_t) directly instead of int (int seems too conservative for size), without an alias in this case

ktf replied:
We should also not have `using sizeType int` in the `o2::Base` namespace, but it should really be a property of the class itself (if really needed, I would simply use size_t).

mkrzewic replied:
size_t is not guaranteed to be the same length on all architectures AFAIK, we  need a fixed size here since it goes into a transport buffer...

ktf replied:
Fine, use int64_t or whatever, I agree. My actual point is that the `using` statement should really be a class level, not namespace level.

mkrzewic replied:
indeed

shahor02 replied:
Ok, changed to ``using sizeType = std::int64_t;`` though hardly int capacity will be insufficient in the class scope

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+   */
+  
+  {
+    std::unique_ptr<char[]> dum(mPtr.release()); // destroy old buffer if any (== delete[] mPtr.release())
```
mPtr.reset(nullptr); would be cleaner if you really need to destroy the resource

shahor02 replied:
thanks, done

mkrzewic replied:
actually I think this should not be here - the internal state should be unchanged after throw - and you might throw just after - so you would both delete the old resource AND leak the new one (since in general you don't know where the exception will be caught).
So i'd remove the "delete" here alltogether. When you change to pass-by-unique pointer you are exception safe.

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+    }
+  } catch(const char* msg) {
+    LOG(FATAL) << msg << " rawPtr is null" << FairLogger::endl;
+  }
```
why exceptions here? it looks like it boils down to
if(rawPtr==nullptr) {LOG(...);}

ktf replied:
Actually, since it's a container class, IMHO the right thing to do is to let it throw and then catch the exception in the user code.

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+  ContVec(char* rawptr, bool copy, sizeType nbytes=-1)  : mPtr(nullptr) {recreate(rawptr,copy,nbytes);}
+
+  // recreate container from received raw pointer on the existing buffer
+  void  recreate(char* rawptr, bool copy, sizeType nbytes=-1);
```
I think a better name would be "adopt" - since you would be adopting an existing buffer. Another issue is the ownership - you take ownership of a raw pointer here for a resource which might be (and usually is) owned by the transport. If you want to transfer ownership take a unique_ptr as argument. We should think about what to do here, maybe it would be better to take the message (FairMQMessage) directly which is a movable owning thing or come up with a different general API/agreement on how to do buffer ownership transfers in and (especially) out of transport.

shahor02 replied:
Right now the ownership is an option: if requested (copy==true), the vector will manage the copy of the provided buffer, otherwise it will take the ownership. The final behaviour should depend on who will actually manage the buffer of FairMGMessage.

mkrzewic replied:
yes, but right now it takes ownership via a raw pointer in the method signature - this is dangerous and does not express ownership semantics. unique_ptr<char[]> should be passed here.

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+ public:
+  
+  struct Header {
+    H         userInfo;         // user assigned data info
```
maybe the userInfo could/should go to the metadata and not directly into the payload buffer?

shahor02 replied:
I thought about H as a short descriptor of the container one detector may pass to another (e.g. struct with some hashes, bitfields etc.) w/o any cost. If we decide that all such metadata are sent by separate message, then I would remove the H completely. Let's keep it until such decision is taken.

mkrzewic commented on Detectors/Base/test/testContVec.cxx:
```diff
@@ -0,0 +1,247 @@
+  TFile* flout = TFile::Open(outRoot.data(),"recreate");
+  cnt.Write("cntw");
+  flout->Close();
+  delete flout;
```
how about something without deletes etc.: 

```
auto flout = make_unique<TFile>(outRoot.data(),"recreate");
cnt.Write("cntw");
flout->Close();
```

shahor02 replied:
ok

mkrzewic commented on Detectors/Base/test/testContVec.cxx:
```diff
@@ -0,0 +1,247 @@
+  nread = 0;
+  ifstream file(fname, ios::in|ios::binary|ios::ate);
+  nread = (int)file.tellg();
+  unique_ptr<char[]> buff(new char[nread]);
```
please use make_unique:

auto buff = make_unique<byte[]>(nread);

shahor02 replied:
done

mkrzewic commented:
Hi @shahor02 , I think this is a very good and needed start - for some more specific comments see the code section.
In general I think we should think about what we want to do with buffer ownership, I think in general what you have in there (except for some bugs in the interface) is OK, me might want to consider adding something that ties more closely with the transport, not sure yet...

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+using sizeType = int;
+
+ template <class T, class H>
+  class ContVec: public TObject {
```
PErhaps the name could be more descriptive? How about just "Vector"?

shahor02 replied:
done

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+
+  
+  std::unique_ptr<char[]> mPtr;                //! pointer on continuos block containing full object data
+  std::unique_ptr<std::vector<T>> mVecForTree; //! vector for writing objects into the tree
```
I think there is no need for wrapping the vector in a unique_ptr, vector already is a movable type...

shahor02 replied:
The mVecForTree is used only for output to TTree. If the latter is not requested, the mVecForTree need not to be initialized. So the alternative to unique_ptr will be the raw pointer (with new/delete). Did I misunderstand you?

mkrzewic replied:
no I meant just std::vector - when untouched it is something like 24bytes (vs 8 bytes of unique_ptr). std::vector will only start allocating stuff when you ask it to and it manages its resources automatically just like unique_ptr - so usually there is no need to wrap it.

shahor02 replied:
ok, done, -16 bytes in our resources...

mkrzewic commented:
actually thinking about it some more I think we should probably define the requirements and interface of a generic vector type like this. One thing which come to mind is that is has to tie in to the transport allocation strategy - only the transport knows how and where to allocate to avoid copying (shared memory etc.). So the type needs to be aware of which transport channel the buffer is destined to go to and use the channel allocator.
So I think this class (ContVec) could be a wrapper of this generic vector class (which would then replace the mPtr member) and only provide the ROOT TTree interface etc.

mkrzewic commented:
I suggest to discuss this next week at the data model meeting.

mkrzewic commented on Detectors/Base/include/DetectorsBase/ContVec.h:
```diff
@@ -0,0 +1,362 @@
+  }
+  
+  if (copy) {
+    mPtr.reset(new char[nbDecoded]());
```
mPtr = make_unique<byte[]>(nbDecoded);

shahor02 committed:
renamed to Vector + rectified according to PR review

shahor02 commented:
Hi, for the ``adopt`` and corresponding c-tor: I split it to pair of methods: one takes as an argument a raw pointer and creates a copy of the buffer. Alternative one takes unique_ptr and always moves the ownership to itself.
The class was renamed to Vector accounting for all comments except the 
1) exception handling (need to think how to do this)
2) global question of @mkrzewic concerning the ownership and allocation: the container is meant not only for the transfer but also to create objects (tracks, vectors etc) for internal use of the component. So, it is not for granted that the transport layer should always own the buffer. Let's discuss this at the meeting.

matthiasrichter commented:
to couple this to the transport sounds weird to me. This would be functionality outside the scope of such a container. It has to support moving the buffers in a generic way,but independent from any transport

mkrzewic commented:
@matthiasrichter the only reason to have a class like this is to avoid a copy on send/receive - if we allow that then using std::vector would be enough. If your target channel for the data is for example  configured to use shared memory and the allocator you are using is the standard heap allocator (say, malloc), you need to copy your data from the heap to shared memory. There is no way to "move" a buffer if it is allocated in a memory segment that is "wrong" for the transport.

mkrzewic commented on Detectors/Base/include/DetectorsBase/Vector.h:
```diff
@@ -0,0 +1,393 @@
+
+//-------------------------------------------------------------------
+template <class T, class H>
+  void Vector<T,H>::clear(bool calldestructor)
```
I think "calldestructor" is unnecessary and it reminds me too much of ROOT IsOwner()-like semantics - which go against a clearly defined ownership model. Like in std::vector you either always destroy if you contain objects directly, or never it you have pointers. I think here we by construction never have to/want to destroy:
by construction T is_trivially_copyable which (I think at least in our case) implies that the dtor has no side effects (since you can just memcpy it and T cannot have pointer members etc.) - so calling the dtor should never be needed.

shahor02 replied:
Hi @mkrzewic , the reason I used destructor in clear (and constructor in push_back instead of simple copy) is that the trivially-copiable does not necessarily mean POD (and even POD is formally not forbidden to have pointers). While the objects used with such with container (apart from being formally trivially-copyable) are meant to have only non-pointer data members, I did not find any generic way to impose that on compilation level (no traits like has_pointers). If you
know such trick, than I'll remove both d-rot in the clear and c-tor in push_back.

mkrzewic replied:
Hi @shahor02 , no I don't, but it does not mean there is none :)
I think even if there is no way of enforcing exactly what we need at compile time the rest will have to be covered by a contract - the data model. As for trivially copyable: when I look at [definition](http://en.cppreference.com/w/cpp/types/is_trivially_copyable) it says:

> Objects of trivially-copyable types are the only C++ objects that may be safely copied with std::memcpy or serialized to/from binary files with std::ofstream::write()/std::ifstream::read()

which I think implies no pointer data members as files/streams are used to communicate between processes - so pointers would not work.

shahor02 replied:
I think these C++ definitions don't assume transfer between different processes, i.e. change of the address space. There is no neither formal nor effective interdiction on pointers. Here is a simple test:
 ````
root [0] std::is_trivially_copyable<TObject>::value 
(const bool) false  //correct 
root [1] struct XX {char* mPtr; void setPtr(char* ptr){mPtr=ptr;}};
root [2] std::is_trivially_copyable<XX>::value
(const bool) true
````

mkrzewic replied:
@shahor02 , that is worrying, apparently only reference types are forbidden (which for some reason does not include pointers, only references). But even then a struct with a reference member is_trivially_copyable, but is not POD and is not standard_layout.
Maybe somebody finds a way, otherwise we'd have to rely on the code checker? or just convention...

shahor02 commented:
Hi @mkrzewic, to me the reason to have such container is not avoiding the copy but avoiding serialization for the transferred data. I don't see how one can avoid at least one copy (or allocating extra buffers) when transferring objects from one component to another.

mkrzewic commented:
Hi @shahor02 , serialization is also a copy. And the copy that I think you mean can be avoided by a fitting allocating strategy (allocate in the proper mem segment, like shared mem) and construct your data directly there. But this we actually could do with std::vector.
The main reason for this container (in my view) is that std::vector does not allow us to adopt it's contents from a buffer - not without the copy step (at least not easily). So it is in this case the zero copy receive that is the main motivation, otherwise we could just use std::vector, why reinvent the wheel. Another, slightly less annoying point is that std::vector always does things like call dtors and ctors etc. which we also prefer to have under our control.

shahor02 committed:
do not call constructors/destructors in push_back/release + cosmetics
adopt(const* char...) renamed to recreate... to emphasize that external buffer is copied

matthiasrichter reviewed (changes requested):
I see the potential of such a class to provide high-level interface (people are used to) at the same time as supporting an optimized memory model. I'm unsure though whether it's smart to re-write exiting (and efficient) standard code. Why not using std containers underneath and adding the high level functionality and the ROOT interface in a wrapper style?

I think we should take this nice initiative and define clear deliverables for such a tool class.

Some more comments in the code, I see that there have been already many in the first version which are not displayed in the updated on, sorry if there might be some duplicates.

matthiasrichter commented:
Probably we should define what we expect in terms of memory model from such a class. First of all, according to Ruben it's supposed to allow exchange of information without serialization. So there is no additional processing needed when shipping data between devices.

At send, the allocator can be chosen according to the needs of the underlying memory model, it's the allocator which couples the thing to e.g. transport, nothing in the class itself. (probably just some wording). That is why std containers provide allocators.

At receive, I agree it's desirable to provide the same container interface instead of accessing an array of objects. Can we do a trick with the allocator here as well?

Using existing std containers would allow us to focus on the memory model instead of maintaining custom code and handling corner cases.

shahor02 commented on Detectors/Base/include/DetectorsBase/Vector.h:
```diff
@@ -0,0 +1,395 @@
+namespace Base {
+
+ template <class T, class H>
+   class Vector: public TObject {
```
I made it from TObject in order to generate automatically Write method (for dumping to root files)

sawenzel replied:
The alternative here is
`TFile f; f.WriteObject()` which works without the Write method from TObject;

sawenzel replied:
I think the name "Vector" for this class is a bit too unspecific and easily confusable with std::vector etc. Can we consider having a more specialized name?

shahor02 replied:
@sawenzel : for the TObject: indeed, thanks for the suggestion, fix pushed. For the name: originally it was vectCont, was changed to Vector acc. reviewer suggestion :) Let's at the moment discuss more global problems, the cosmetics can be done later.

mkrzewic commented:
@matthiasrichter I agree it would be preferable to use std::XXX but as I mentioned somewhere in the thread there is no trivial way to adopt a buffer. I think maybe it could be possible via some allocator_traits magic on an intermediate type which is movable into a vector - but I did not look into details yet. Even is such a trick would be possible, question is: would it be sustainable...

shahor02 committed:
just throw the exception, catch will go to user code

shahor02 commented on Detectors/Base/include/DetectorsBase/Vector.h:
```diff
@@ -0,0 +1,395 @@
+   * nbytes: expected size of the buffer in bytes, used for error check
+   */
+  
+  try {
```
done in my last commit

ironMann commented:
@mkrzewic well, std::containers are [allocator-aware since c++11](http://en.cppreference.com/w/cpp/concept/AllocatorAwareContainer). You only need a custom per-channel-type allocator class, a fancy pointer class (it's really called like that) which can encode raw, shm and rdma memory pointer, and you're golden :)

shahor02 committed:
No TObject inheritance, use TFile::WriteObject to store

shahor02 commented:
@ironMann : could you show a snippet which would do the job?

ironMann commented:
@shahor02 Here are some references to back up my last statement: A good starting tutorial, although not complete, would be [this](https://rawgit.com/google/cxx-std-draft/allocator-paper/allocator_user_guide.html). Concerning shm, boost has snippets for [map](http://www.boost.org/doc/libs/1_63_0/doc/html/interprocess/quick_guide.html#interprocess.quick_guide.qg_interprocess_map) and [vector](http://www.boost.org/doc/libs/1_63_0/doc/html/interprocess/quick_guide.html#interprocess.quick_guide.qg_interprocess_container) using [offset pointers](http://www.boost.org/doc/libs/1_63_0/doc/html/interprocess/quick_guide.html#interprocess.quick_guide.qg_offset_ptr).

mkrzewic commented:
@ironMann indeed, boost::interprocess::vector might be an option. We would, however, need a standards compliant allocator API from fairmq. We should look into it.
So we are not talking about std::containers here - they only can do half a job as mentioned earlier in the thread - AFAIK there is no way to "read" or "populate" them from a buffer without making a copy.

mkrzewic commented:
@ironMann actually, having a brief look, boost::interprocess:vector would also not solve the issue: it actually does the same as std::vector: it is not contiguous as it involves 2 separate allocations: one for the vector holding a (offset) handle to the store, one for the backing store itself (the array). So it only works in a true shared memory segment, would not work across an arbitrary transport as it also does not seem to have an "adopt a buffer" functionality...

sawenzel commented:
Reading the other comments, I think I would call this class `InterprocessVector` or `IPCVector` or something like this.

ktf commented on Detectors/Base/include/DetectorsBase/Vector.h:
```diff
@@ -0,0 +1,384 @@
+  if (ptr==nullptr) {
+    std::ostringstream strErr;
+    strErr << "invalid arguments: ptr is null" << std::endl;
+    throw strErr.str();
```
Why not simply `throw std::runtime_error("invalid arguments: ptr is null")`?

shahor02 replied:
because I never used exceptions before, thanks for the suggestion, changed.

ktf commented on Detectors/Base/include/DetectorsBase/Vector.h:
```diff
@@ -0,0 +1,384 @@
+    std::ostringstream strErr;
+    strErr << "invalid arguments: " << "wrong size " << nbytes <<
+      " at least " << dataOffset() << " expected" << std::endl;
+    throw  strErr.str();
```
I would throw a `std::runtime_error(strErr.str())` rather than a string.

shahor02 committed:
throw std::runtime_error rather than the message

alibuild commented:
Error while checking build/O2/o2 for af993a31041ec8e4817f32f5eae79ac5fb80ef34:
```
sw/BUILD/O2-latest/log
+ ln -sf /mnt/mesos/sandbox/sandbox/sw/BUILD/c814e78a67cf4e92a76e48b3e33b06e4d4eade7d/O2/compile_commands.json /mnt/mesos/sandbox/sandbox/O2/compile_commands.json
+ [[ -n 1 ]]
+ make test
Running tests...
Test project /mnt/mesos/sandbox/sandbox/sw/BUILD/c814e78a67cf4e92a76e48b3e33b06e4d4eade7d/O2
      Start  1: testMessageList
 1/18 Test  #1: testMessageList ..................   Passed    0.00 sec
      Start  2: testDataHeader
 2/18 Test  #2: testDataHeader ...................   Passed    0.49 sec
      Start  3: testTimeStamp
 3/18 Test  #3: testTimeStamp ....................   Passed    0.51 sec
      Start  4: testBasicHits
 4/18 Test  #4: testBasicHits ....................   Passed    0.57 sec
      Start  5: TimeFrameTest
 5/18 Test  #5: TimeFrameTest ....................   Passed    0.61 sec
      Start  6: testVector
 6/18 Test  #6: testVector .......................   Passed    0.68 sec
      Start  7: testTPCBase
 7/18 Test  #7: testTPCBase ......................   Passed    0.47 sec
      Start  8: testTPCCalDet
 8/18 Test  #8: testTPCCalDet ....................   Passed    0.87 sec
      Start  9: testTPCMapper
 9/18 Test  #9: testTPCMapper ....................   Passed    0.49 sec
      Start 10: testTPCSyncPatternMonitor
10/18 Test #10: testTPCSyncPatternMonitor ........   Passed    0.58 sec
      Start 11: testTPCAdcClockMonitor
11/18 Test #11: testTPCAdcClockMonitor ...........   Passed    0.47 sec
      Start 12: testTPCElectronTransport
12/18 Test #12: testTPCElectronTransport .........   Passed    0.65 sec
      Start 13: testTPCGEMAmplification
13/18 Test #13: testTPCGEMAmplification ..........   Passed   27.48 sec
      Start 14: testTPCSimulation
14/18 Test #14: testTPCSimulation ................   Passed    0.44 sec
      Start 15: testExampleModule1
15/18 Test #15: testExampleModule1 ...............   Passed    0.01 sec
      Start 16: testMessageFormat
16/18 Test #16: testMessageFormat ................***Failed    0.50 sec
      Start 17: O2MessageMonitorTest
17/18 Test #17: O2MessageMonitorTest .............   Passed    0.43 sec
      Start 18: test_TimeframeParser
18/18 Test #18: test_TimeframeParser .............   Passed    0.42 sec

94% tests passed, 1 tests failed out of 18

Total Test time (real) =  35.70 sec

The following tests FAILED:
	 16 - testMessageFormat (Failed)
Errors while running CTest
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/af993a31041ec8e4817f32f5eae79ac5fb80ef34/fullLog.txt).

alibuild commented:
Error while checking build/o2checkcode/o2-daq for af993a31041ec8e4817f32f5eae79ac5fb80ef34:
```
sw/BUILD/O2-latest/log
+ cp /mnt/mesos/sandbox/sandbox/sw/BUILD/eff886d1b3909e0087cb93e66755cb6a3fab6f7e/O2/compile_commands.json /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/O2/0_O2_DAQ-1
++ readlink /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0
+ DEVEL_SOURCES=/mnt/mesos/sandbox/sandbox/O2
+ '[' /mnt/mesos/sandbox/sandbox/O2 '!=' /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0 ']'
+ perl -p -i -e 's|/mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/0_O2_DAQ/0|/mnt/mesos/sandbox/sandbox/O2|' compile_commands.json
+ ln -sf /mnt/mesos/sandbox/sandbox/sw/BUILD/eff886d1b3909e0087cb93e66755cb6a3fab6f7e/O2/compile_commands.json /mnt/mesos/sandbox/sandbox/O2/compile_commands.json
+ [[ -n 1 ]]
+ make test
Running tests...
Test project /mnt/mesos/sandbox/sandbox/sw/BUILD/eff886d1b3909e0087cb93e66755cb6a3fab6f7e/O2
      Start  1: testMessageList
 1/18 Test  #1: testMessageList ..................   Passed    0.01 sec
      Start  2: testDataHeader
 2/18 Test  #2: testDataHeader ...................   Passed    1.29 sec
      Start  3: testTimeStamp
 3/18 Test  #3: testTimeStamp ....................   Passed    0.49 sec
      Start  4: testBasicHits
 4/18 Test  #4: testBasicHits ....................   Passed    0.53 sec
      Start  5: TimeFrameTest
 5/18 Test  #5: TimeFrameTest ....................   Passed    0.53 sec
      Start  6: testVector
 6/18 Test  #6: testVector .......................   Passed    0.81 sec
      Start  7: testTPCBase
 7/18 Test  #7: testTPCBase ......................   Passed    0.47 sec
      Start  8: testTPCCalDet
 8/18 Test  #8: testTPCCalDet ....................   Passed    0.87 sec
      Start  9: testTPCMapper
 9/18 Test  #9: testTPCMapper ....................   Passed    0.50 sec
      Start 10: testTPCSyncPatternMonitor
10/18 Test #10: testTPCSyncPatternMonitor ........   Passed    0.53 sec
      Start 11: testTPCAdcClockMonitor
11/18 Test #11: testTPCAdcClockMonitor ...........   Passed    0.58 sec
      Start 12: testTPCElectronTransport
12/18 Test #12: testTPCElectronTransport .........   Passed    0.79 sec
      Start 13: testTPCGEMAmplification
13/18 Test #13: testTPCGEMAmplification ..........   Passed   25.21 sec
      Start 14: testTPCSimulation
14/18 Test #14: testTPCSimulation ................   Passed    0.50 sec
      Start 15: testExampleModule1
15/18 Test #15: testExampleModule1 ...............   Passed    0.02 sec
      Start 16: testMessageFormat
16/18 Test #16: testMessageFormat ................   Passed    0.45 sec
      Start 17: O2MessageMonitorTest
17/18 Test #17: O2MessageMonitorTest .............   Passed    0.48 sec
      Start 18: test_TimeframeParser
18/18 Test #18: test_TimeframeParser .............   Passed    0.54 sec

100% tests passed, 0 tests failed out of 18

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/af993a31041ec8e4817f32f5eae79ac5fb80ef34/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2 for af993a31041ec8e4817f32f5eae79ac5fb80ef34:
```

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/af993a31041ec8e4817f32f5eae79ac5fb80ef34/fullLog.txt).

alibuild commented:
Error while checking build/O2/o2-dev-fairroot for af993a31041ec8e4817f32f5eae79ac5fb80ef34:
```

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/af993a31041ec8e4817f32f5eae79ac5fb80ef34/fullLog.txt).

shahor02 committed:
Fixed copyright statement

shahor02 commented:
Hello,
will  this PR be eventually merged? Any extra comments apart from renaming the class to IPCVector
(i.e. Inter-Process-Communiation vector as @sawenzel suggested)?

alibuild commented:
Error while checking build/o2checkcode/o2-daq for 210ac5625b2a00400b4709a9b29c47e0bc5081ee:
```
sw/BUILD/O2-latest/log
      Start  3: testTimeStamp
 3/25 Test  #3: testTimeStamp ....................   Passed    0.43 sec
      Start  4: testBasicHits
 4/25 Test  #4: testBasicHits ....................   Passed    0.46 sec
      Start  5: TimeFrameTest
 5/25 Test  #5: TimeFrameTest ....................   Passed    0.57 sec
      Start  6: testVector
 6/25 Test  #6: testVector .......................   Passed    0.68 sec
      Start  7: testTPCBase
 7/25 Test  #7: testTPCBase ......................   Passed    0.51 sec
      Start  8: testTPCCalDet
 8/25 Test  #8: testTPCCalDet ....................   Passed    0.78 sec
      Start  9: testTPCMapper
 9/25 Test  #9: testTPCMapper ....................   Passed    0.46 sec
      Start 10: testTPCSyncPatternMonitor
10/25 Test #10: testTPCSyncPatternMonitor ........   Passed    0.46 sec
      Start 11: testTPCAdcClockMonitor
11/25 Test #11: testTPCAdcClockMonitor ...........   Passed    0.78 sec
      Start 12: testTPCElectronTransport
12/25 Test #12: testTPCElectronTransport .........   Passed    0.64 sec
      Start 13: testTPCGEMAmplification
13/25 Test #13: testTPCGEMAmplification ..........   Passed   24.57 sec
      Start 14: testTPCSimulation
14/25 Test #14: testTPCSimulation ................   Passed    0.39 sec
      Start 15: testExampleModule1
15/25 Test #15: testExampleModule1 ...............   Passed    0.01 sec
      Start 16: testMessageFormat
16/25 Test #16: testMessageFormat ................   Passed    0.52 sec
      Start 17: O2MessageMonitorTest
17/25 Test #17: O2MessageMonitorTest .............   Passed    0.39 sec
      Start 18: test_TimeframeParser
18/25 Test #18: test_TimeframeParser .............   Passed    0.38 sec
      Start 19: test_SubframeUtils01
19/25 Test #19: test_SubframeUtils01 .............   Passed    0.38 sec
      Start 20: test_PayloadMerger01
20/25 Test #20: test_PayloadMerger01 .............   Passed    0.38 sec
      Start 21: test_dc_primitives
21/25 Test #21: test_dc_primitives ...............   Passed    0.01 sec
      Start 22: test_Fifo
22/25 Test #22: test_Fifo ........................   Passed    7.01 sec
      Start 23: test_DataGenerator
23/25 Test #23: test_DataGenerator ...............   Passed    1.36 sec
      Start 24: test_HuffmanCodec
24/25 Test #24: test_HuffmanCodec ................   Passed    0.45 sec
      Start 25: test_DataDeflater
25/25 Test #25: test_DataDeflater ................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 25

```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/389/210ac5625b2a00400b4709a9b29c47e0bc5081ee/fullLog.txt).

shahor02 committed:
Merge branch 'dev' into dev_cont
